// @prisma/ts-builders/dist/Export.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Export_exports = {};
__export(Export_exports, {
  Export: () => Export,
  moduleExport: () => moduleExport
});
module.exports = __toCommonJS(Export_exports);
class Export {
  constructor(declaration) {
    this.declaration = declaration;
  }
  docComment;
  setDocComment(docComment) {
    this.docComment = docComment;
    return this;
  }
  write(writer) {
    if (this.docComment) {
      writer.write(this.docComment);
    }
    writer.write("export ").write(this.declaration);
  }
}
function moduleExport(declaration) {
  return new Export(declaration);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Export,
  moduleExport
});
