// @prisma/ts-builders/dist/ExportFrom.mjs
class NamespaceExport {
  constructor(from, namespace) {
    this.from = from;
    this.namespace = namespace;
  }
  write(writer) {
    writer.write(`export * as ${this.namespace} from '${this.from}'`);
  }
}
class BindingsExport {
  constructor(from) {
    this.from = from;
  }
  namedExports = [];
  named(namedExport2) {
    if (typeof namedExport2 === "string") {
      namedExport2 = new NamedExport(namedExport2);
    }
    this.namedExports.push(namedExport2);
    return this;
  }
  write(writer) {
    writer.write("export ").write("{ ").writeJoined(", ", this.namedExports).write(" }").write(` from "${this.from}"`);
  }
}
class NamedExport {
  constructor(name) {
    this.name = name;
  }
  alias;
  type = false;
  as(alias) {
    this.alias = alias;
    return this;
  }
  typeOnly() {
    this.type = true;
    return this;
  }
  write(writer) {
    if (this.type) {
      writer.write("type ");
    }
    writer.write(this.name);
    if (this.alias) {
      writer.write(" as ").write(this.alias);
    }
  }
}
class ExportAllFrom {
  constructor(from) {
    this.from = from;
  }
  asNamespace(namespace) {
    return new NamespaceExport(this.from, namespace);
  }
  named(binding) {
    return new BindingsExport(this.from).named(binding);
  }
  write(writer) {
    writer.write(`export * from "${this.from}"`);
  }
}
function moduleExportFrom(from) {
  return new ExportAllFrom(from);
}
function namedExport(name) {
  return new NamedExport(name);
}
export {
  BindingsExport,
  ExportAllFrom,
  NamedExport,
  NamespaceExport,
  moduleExportFrom,
  namedExport
};
