// @prisma/ts-builders/dist/StringLiteralType.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var StringLiteralType_exports = {};
__export(StringLiteralType_exports, {
  StringLiteralType: () => StringLiteralType,
  StringLiteralValue: () => StringLiteralValue,
  stringLiteral: () => stringLiteral
});
module.exports = __toCommonJS(StringLiteralType_exports);
var import_TypeBuilder = require("./TypeBuilder");
var import_ValueBuilder = require("./ValueBuilder");
class StringLiteralType extends import_TypeBuilder.TypeBuilder {
  constructor(content) {
    super();
    this.content = content;
  }
  write(writer) {
    writer.write(JSON.stringify(this.content));
  }
  asValue() {
    return new StringLiteralValue(this);
  }
}
class StringLiteralValue extends import_ValueBuilder.ValueBuilder {
  #type;
  constructor(type) {
    super();
    this.#type = type;
  }
  write(writer) {
    writer.write(this.#type);
  }
}
function stringLiteral(content) {
  return new StringLiteralType(content);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  StringLiteralType,
  StringLiteralValue,
  stringLiteral
});
